File: backend/src/shared/metrics/metrics.ts
/**
 * Minimal, dependency-free in-process metrics registry (P7.3).
 *
 * Deliberately not a Prometheus client: no scrape target (Prometheus/
 * Grafana) exists anywhere in this repository or environment, so pulling in
 * `prom-client` for a text-exposition format nothing consumes would be an
 * unused dependency. This exposes the same underlying counts as a small
 * JSON summary instead (`GET /metrics`), which is enough to answer "how
 * many requests, how fast, how many blockchain health checks failed" —
 * the actual questions this phase's brief asks for — without inventing
 * infrastructure this project doesn't otherwise have.
 */

interface RouteStats {
  count: number;
  totalDurationMs: number;
  statusCounts: Record<string, number>;
}

interface BlockchainStats {
  checks: number;
  connected: number;
  unavailable: number;
  proxyUnreachable: number;
}

export interface MetricsRegistry {
  recordRequest(method: string, route: string, statusCode: number, durationMs: number): void;
  recordBlockchainCheck(status: string): void;
  snapshot(): MetricsSnapshot;
  reset(): void;
}

export interface MetricsSnapshot {
  uptimeSeconds: number;
  requests: {
    total: number;
    byRoute: Array<{
      method: string;
      route: string;
      count: number;
      avgDurationMs: number;
      statusCounts: Record<string, number>;
    }>;
  };
  blockchain: BlockchainStats;
}

/** Creates a fresh, process-local metrics registry. Not persisted, not shared across instances. */
export function createMetricsRegistry(now: () => number = Date.now): MetricsRegistry {
  const routeStats = new Map<string, RouteStats>();
  const blockchain: BlockchainStats = {
    checks: 0,
    connected: 0,
    unavailable: 0,
    proxyUnreachable: 0,
  };
  const startedAt = now();

  return {
    recordRequest(method, route, statusCode, durationMs): void {
      const key = `${method} ${route}`;
      const existing = routeStats.get(key) ?? { count: 0, totalDurationMs: 0, statusCounts: {} };
      existing.count += 1;
      existing.totalDurationMs += durationMs;
      const statusKey = String(statusCode);
      existing.statusCounts[statusKey] = (existing.statusCounts[statusKey] ?? 0) + 1;
      routeStats.set(key, existing);
    },

    recordBlockchainCheck(status): void {
      blockchain.checks += 1;
      if (status === 'connected') {
        blockchain.connected += 1;
      } else if (status === 'proxy_unreachable') {
        blockchain.proxyUnreachable += 1;
      } else {
        blockchain.unavailable += 1;
      }
    },

    snapshot(): MetricsSnapshot {
      const byRoute = [...routeStats.entries()].map(([key, stats]) => {
        const [method, ...routeParts] = key.split(' ');
        return {
          method: method ?? '',
          route: routeParts.join(' '),
          count: stats.count,
          avgDurationMs: stats.count > 0 ? stats.totalDurationMs / stats.count : 0,
          statusCounts: { ...stats.statusCounts },
        };
      });
      const total = byRoute.reduce((sum, r) => sum + r.count, 0);

      return {
        uptimeSeconds: (now() - startedAt) / 1000,
        requests: { total, byRoute },
        blockchain: { ...blockchain },
      };
    },

    reset(): void {
      routeStats.clear();
      blockchain.checks = 0;
      blockchain.connected = 0;
      blockchain.unavailable = 0;
      blockchain.proxyUnreachable = 0;
    },
  };
}

File: intelligence/device_ai/utils/metrics.py
"""Minimal, dependency-free in-process metrics registry (P7.3).

Mirrors ``backend/src/shared/metrics/metrics.ts``: no Prometheus client is
introduced here either, for the same reason — no scrape target exists
anywhere in this repository or environment, so a text-exposition format
nothing consumes would be an unused dependency. ``GET /metrics`` exposes the
same underlying counts as a small JSON summary instead.
"""

from __future__ import annotations

import time
from collections.abc import Callable
from dataclasses import dataclass, field
from threading import Lock


@dataclass
class _RouteStats:
    count: int = 0
    total_duration_ms: float = 0.0
    status_counts: dict[str, int] = field(default_factory=dict)


@dataclass
class _FabricStats:
    transactions: int = 0
    succeeded: int = 0
    failed: int = 0


class MetricsRegistry:
    """Process-local metrics registry. Not persisted, not shared across processes."""

    def __init__(self, *, clock: Callable[[], float] = time.monotonic) -> None:
        self._clock = clock
        self._started_at = clock()
        self._lock = Lock()
        self._routes: dict[tuple[str, str], _RouteStats] = {}
        self._fabric = _FabricStats()

    def record_request(
        self, method: str, path: str, status_code: int, duration_ms: float
    ) -> None:
        """Record one completed HTTP request against its matched route."""
        key = (method, path)
        with self._lock:
            stats = self._routes.setdefault(key, _RouteStats())
            stats.count += 1
            stats.total_duration_ms += duration_ms
            status_key = str(status_code)
            stats.status_counts[status_key] = stats.status_counts.get(status_key, 0) + 1

    def record_fabric_transaction(self, *, succeeded: bool) -> None:
        """Record one completed Fabric Gateway transaction attempt."""
        with self._lock:
            self._fabric.transactions += 1
            if succeeded:
                self._fabric.succeeded += 1
            else:
                self._fabric.failed += 1

    def snapshot(self) -> dict:
        """Return a point-in-time JSON-serializable copy of all recorded metrics."""
        with self._lock:
            by_route = [
                {
                    "method": method,
                    "route": path,
                    "count": stats.count,
                    "avg_duration_ms": (
                        stats.total_duration_ms / stats.count
                        if stats.count > 0
                        else 0.0
                    ),
                    "status_counts": dict(stats.status_counts),
                }
                for (method, path), stats in self._routes.items()
            ]
            total = sum(stats.count for stats in self._routes.values())
            fabric = {
                "transactions": self._fabric.transactions,
                "succeeded": self._fabric.succeeded,
                "failed": self._fabric.failed,
            }

        return {
            "uptime_seconds": self._clock() - self._started_at,
            "requests": {"total": total, "by_route": by_route},
            "fabric": fabric,
        }

    def reset(self) -> None:
        """Clear all recorded data. Intended for test isolation."""
        with self._lock:
            self._routes.clear()
            self._fabric = _FabricStats()


_registry = MetricsRegistry()


def get_metrics_registry() -> MetricsRegistry:
    """Return the process-wide metrics registry singleton."""
    return _registry
